packages/core/src/impl-store.ts

```diff
@@ -41,15 +45,24 @@ async function fetchOrDeployGeneric<T extends Deployment>(
       debug('fetching deployment of', lens.description);
       const data = await manifest.read();
       const deployment = lens(data);
-      const stored = deployment.get();
-      if (stored === undefined) {
-        debug('deployment of', lens.description, 'not found');
-      }
-      const updated = await resumeOrDeploy(provider, stored, deploy);
-      if (updated !== stored) {
+      let updated;
+      const stored = await getAndValidate<T>(deployment, lens, provider);
+      if (merge) {
+        updated = await deploy();
         await checkForAddressClash(provider, data, updated);
-        deployment.set(updated);
+        if (deployment.merge) {
+          await deployment.merge(updated);
+        } else {
```

Comment:
It seems to me that we should only allow `merge = true` if `deployment.merge` exists, would you say this is right? If so, the else branch shouldn't be here and we should error merge is set incorrectly.

Reply:
Changed it to throw an error [here](https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/517/commits/57e110956fdcda903939d85ee9ee7d60b3ddcdc3#diff-74a100ab37e69fa1bd58695b1f5b6870c10f5a7c542ef38eb19e3ebe7a3d195eR47), but the TS compiler gave an error that `deployment.merge` may be undefined [here](https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/517/commits/57e110956fdcda903939d85ee9ee7d60b3ddcdc3#diff-74a100ab37e69fa1bd58695b1f5b6870c10f5a7c542ef38eb19e3ebe7a3d195eR58) so I had to force it with `deployment.merge!(updated);`.  This leads to a linting warning `Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion`

I could put the check directly before calling `deployment.merge(updated);` but then this could be after an actual deployment occurred.  It is better to check it earlier.  Any thoughts?

Reply:
The simplest fix is to write `if (merge && deployment.merge)` even if we know the first condition implies the second.